### ISSUE TITLE
how to trim leading and tailing blanks in lines separated by crlf \r\n

### ISSUE THREAD

**Issue by vsrawat** [Woundorf/foxreplace#233]
I need to trim lines, meaning all spaces/ whitespaces at start and end lines are to be removed.

these are not working
{"input": "^\\s+", "inputType": "regexp", "output": "", "caseSensitive": false},
{"input": "\\s+$", "inputType": "regexp", "output": "", "caseSensitive": false},

putting " " (one space) in place of \\s (all kind of whitespace) also doesn't work.
putting {1.} (one or more) in place of + (one or more) also doesn't work.

it seems fr takes and processes only one line (separated by crlf \r\n) at a time.
so it is taking entire input text as a single unit and ^ works at end of entire input only, and $ works at end of entire input.
how to make it work across each line (separated by crlf \r\n)

in summary, how can fr work in what is called multiline mode, where it can work across crlf \r\n at the same time?

it is text input, unicode, no html/ xml.

thanks.

**Comment by SaltatorMortis** [comment]
`(^|[\r\n])\s+|\s+(?=[\n\r]|$)` with an output of `$1` would do it for you

**Comment by Woundorf** [comment]
Thank you @SaltatorMortis.

@vsrawat You can also follow ticket #95 for a possible change on multiline mode in the future.

**Comment by vsrawat** [comment]
Hi Mortis,

Your regex worded perfectly well and removed all leading and trailing 
whitespaces of each line separated with \r\n.

thanks a gig for it.

However, as there were other phrases also in my text only at start of 
line or end of line, I separated the regex to apply at
1. start of line,
2. end of line,

so I broke it into two.
(^|[\\r\\n])\\s+
\\s+(?=[\\n\\r]|$)
and replaced each with $1

first one worked well and removed all whitespaces at start of line.

but the second one did remove all whitespaces at start of line, BUT IT 
ALSO PUT A $1 in place of removed whitespaces.

with my minimal knowledge of regex, I saw the differences in two parts, 
and found that ?= was different so I removed it.

now,
(^|[\\r\\n])\\s+
\\s+([\\n\\r]|$)

are each working well and removing all leading/ trailing whitespaces.

Now I can put my other phrases in place \\s+ in either or both parts and 
those will be removed at start/ end of line.

Thanks again for it.

Could you please describe what was ?= doing in the second part of regex?

Am I loosing some functionality of the regex by removing ?=

Thanks.
-- 
Rawat

**Comment by vsrawat** [comment]
https://stackoverflow.com/questions/47608903/foxreplace-code-to-trim-lines-at-start-end/47842497#47842497

I had asked the same question at stackoverflow, but no reply was found there.

I have now taken freedom to post the above regex there, so that others can utilise it.

Hope Mr Mortis doesn't mind metioning his name there for credit of the answer.

Thanks.

**Comment by vsrawat** [comment]
Wow
the regex also deleted empty lines, with whitespaces or without 
whitespace, (multiple \r\n)

Actually, I was going to ask how to delete multiple crlf, but the regex 
already does it.

It is great,
but just in case, if I ever want to retain all crlfs as such (say just 
to retain sequence for data taken from some ms word table or ms excel 
cells), how to change this regex that
it DOESN't remove even a single \r\n itself, and removes only whitespaces.

Thanks.
-- 
Rawat

**Comment by SaltatorMortis** [comment]
the (?=) is a positive Lookahead
this means it searches for a whitespace followed by a` [\\n\\r]|$` BUT it does it not include into the result.
the code `\\s+(?=[\\n\\r]|$)` does not need the` $1` in the output.
main reason i used this is to keep it a one-liner and the next regex.replace() starts AFTER the end of the replace result (it would capture whitespace at the beginning only if the last line hadn't any whitespace at the end)

to maintain empty lines you need to "encrypt" them first, the reason for this is that \\s contains all Linefeeds
simply replace then the `[\\r\\n]` with your encrypted linefeed and after you're done convert them back to Linefeeds

**Comment by SaltatorMortis** [comment]
hehe dont think regex in the morning, scratch the last part(or use it still valid and usable)
if you want no linefeeds removed replace \s with `[\u0020\u00A0\u1680\u180E\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF\t\v\f]` (any whitespace exept \n\r)
(`[\u0020\u00A0\t\v\f]` for short [space,nbsp,Tab,Vertical tab,Form feed])

**Comment by vsrawat** [comment]
\s includes linefeed...

Do you mean only linefeed (only \n) and no carriage return (without \r)?

If it includes only linefeed (\n), it should have removed only \n and 
would have left \r intact, but that was not so. the regex removed \r\n both.

if it includes only linefeed (\n) and removes only that, then 
pre-encrypting is not required, because \r is still left intact and we 
can then just all replace \r with \r\n.

or you meant it includes both cr lf, in that sequence?

Thanks.
-- 
Rawat

**Comment by vsrawat** [comment]
in my text I have already replace \t with space,
so there is no other whitespace in my text, except space.

I used \s just for eventuality of something else existing there.

I can just use " " (space) in place of \s and the regex would work for 
removing leading trailing spaces and would not remove cr lf.

Thanks.
-- 
Rawat

**Comment by SaltatorMortis** [comment]
\r is inside \s (otherwise it wouldn't remove the emtylines in the sample)
and if you have only space you can search only space jup but if there is some nbsp etc it wouldn't remove it but you can use a simple capture group for this (looks ugly but works)
`[\u0020\u00A0\u1680\u180E\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF\t\v\f]`

a good reference site for Regex is http://www.rexegg.com/

javascripts definition of \s is: ( source xregexp.com )

```
-     U+0009 — Tab — \t
-     U+000A — Line feed — \n
-     U+000B — Vertical tab — \v
-     U+000C — Form feed — \f
-     U+000D — Carriage return — \r
-     U+0020 — Space
-     U+00A0 — No-break space
-     U+1680 — Ogham space mark
-     U+180E — Mongolian vowel separator
-     U+2000 — En quad
-     U+2001 — Em quad
-     U+2002 — En space
-     U+2003 — Em space
-     U+2004 — Three-per-em space
-     U+2005 — Four-per-em space
-     U+2006 — Six-per-em space
-     U+2007 — Figure space
-     U+2008 — Punctuation space
-     U+2009 — Thin space
-     U+200A — Hair space
-     U+2028 — Line separator
-     U+2029 — Paragraph separator
-     U+202F — Narrow no-break space
-     U+205F — Medium mathematical space
-     U+3000 — Ideographic space
-     U+FEFF — Zero width no-break space
```

**Comment by vsrawat** [comment]
if this is the complete list of such possibilities, I have nothing more 
to do than putting this in my json file.

You have prepared it.

You are so thorough, perfection incarnated. :-)

Thanks for resolving it for ever.

Thanks.
-- 
Rawat

**Comment by vsrawat** [comment]
I would add these also.

ZERO WIDTH SPACE zws hex 200B, decimal 8203
ZERO WIDTH NON JOINER zwnj 0wnj hex 200C, decimal 8204
ZERO WIDTH JOINER zwj 0wj hex 200D, decimal 8205

My language is Hindi and such things are often used in Indic languages.

So, if any foxreplace user is referring to this list in future, he would 
find the above ready here.

Thanks.
-- 
Rawat
